src/transformers/models/pix2struct/modeling_pix2struct.py

```diff
@@ -0,0 +1,1686 @@
+
+        # convert into half-precision if necessary
+        if self.weight.dtype in [torch.float16, torch.bfloat16]:
+            hidden_states = hidden_states.to(self.weight.dtype)
+
+        return self.weight * hidden_states
+
+
+try:
+    from apex.normalization import FusedRMSNorm
+
+    Pix2StructLayerNorm = FusedRMSNorm  # noqa
+
+    logger.info("Discovered apex.normalization.FusedRMSNorm - will use it instead of Pix2StructLayerNorm")
+except ImportError:
+    # using the normal Pix2StructLayerNorm
+    pass
+except Exception:
+    logger.warning("discovered apex but it failed to load, falling back to Pix2StructLayerNorm")
+    pass
+
+ALL_LAYERNORM_LAYERS.append(Pix2StructLayerNorm)
+
+
+class Pix2StructVisionEmbeddings(nn.Module):
+    r"""
+    Construct the embeddings from patch. In `Pix2Struct` the input is different from classic Vision-transformer models.
+    Here the input is a sequence of `seq_len` patches. Each patch is represented by a vector of `hidden_size` values.
+    """
+
+    def __init__(self, config: Pix2StructConfig) -> None:
+        super().__init__()
+        self.patch_projection = nn.Linear(config.hidden_size, config.hidden_size)
+
+        self.row_embedder = nn.Embedding(config.seq_len, config.hidden_size)
+        self.column_embedder = nn.Embedding(config.seq_len, config.hidden_size)
+
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+
+    def forward(self, pixel_embeds: torch.Tensor) -> torch.Tensor:
+        # the row and column indices are stored in the first and second position of the pixel_embeds
+        # pixel_embeds: `batch_size`, `seq_len`, `hidden_size` + 2
+        row_indices = pixel_embeds[:, :, 0].long()
+        col_indices = pixel_embeds[:, :, 1].long()
+
+        pixel_embeds = pixel_embeds[:, :, 2:]
+
+        embeddings = self.patch_projection(pixel_embeds)
+        row_embeddings = self.row_embedder(row_indices)
+        col_embeddings = self.column_embedder(col_indices)
+
+        # sum all embeddings together
+        embeddings = embeddings + row_embeddings + col_embeddings
+
+        embeddings = self.dropout(embeddings)
+
+        return embeddings
+
+
+class Pix2StructVisionAttention(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.hidden_size = config.hidden_size
+        self.key_value_proj_dim = config.d_kv
+        self.n_heads = config.num_attention_heads
+        self.dropout = config.attention_dropout
+        self.inner_dim = self.n_heads * self.key_value_proj_dim
+
+        # Mesh TensorFlow initialization to avoid scaling before softmax
+        self.query = nn.Linear(self.hidden_size, self.inner_dim, bias=False)
+        self.key = nn.Linear(self.hidden_size, self.inner_dim, bias=False)
+        self.value = nn.Linear(self.hidden_size, self.inner_dim, bias=False)
+        self.o = nn.Linear(self.inner_dim, self.hidden_size, bias=False)
```

Comment:
if this is not copied from, better variable namesplese (o_

Reply:
Fixed!